autoprecompiles/src/low_degree_bus_interaction_optimizer.rs

```diff
@@ -0,0 +1,562 @@
+
+/// An optimizer that replaces some stateless bus interactions (a.k.a. lookups)
+/// by low-degree algebraic constraints.
+pub struct LowDegreeBusInteractionOptimizer<'a, T, V, S, B> {
+    solver: &'a mut S,
+    bus_interaction_handler: B,
+    degree_bound: DegreeBound,
+    _phantom: PhantomData<(T, V)>,
+}
+
+impl<
+        'a,
+        T: FieldElement,
+        V: Ord + Clone + Ord + Eq + Display + Hash,
+        S: Solver<T, V>,
+        B: BusInteractionHandler<T> + IsBusStateful<T> + RangeConstraintHandler<T>,
+    > LowDegreeBusInteractionOptimizer<'a, T, V, S, B>
+{
+    pub fn new(solver: &'a mut S, bus_interaction_handler: B, degree_bound: DegreeBound) -> Self {
+        Self {
+            solver,
+            bus_interaction_handler,
+            degree_bound,
+            _phantom: PhantomData,
+        }
+    }
+
+    pub fn optimize(self, mut system: ConstraintSystem<T, V>) -> ConstraintSystem<T, V> {
+        let mut new_constraints: Vec<GroupedExpression<T, V>> = vec![];
+        system.bus_interactions = system
+            .bus_interactions
+            .into_iter()
+            .flat_map(|bus_int| {
+                if let Some(replacement) = self.try_replace_bus_interaction(&bus_int) {
+                    // If we found a replacement, add the polynomial constraints and replace
+                    // the bus interaction with interactions implementing the range constraints.
+                    // Note that many of these may be optimized away by the range constraint optimizer.
+                    new_constraints.push(replacement.polynomial_constraint);
+                    self.bus_interaction_handler
+                        .batch_make_range_constraints(replacement.range_constraints)
+                } else {
+                    // Keep the bus interaction as is if a replacement can't be found.
+                    vec![bus_int]
+                }
+            })
+            .collect();
+
+        // Knowing the low-degree functions might help the solver.
+        // The range constraints do not need to be added, because they don't carry information
+        // that is not already implied by the existing bus interactions.
+        self.solver
+            .add_algebraic_constraints(new_constraints.iter().cloned());
+
+        system.algebraic_constraints.extend(new_constraints);
+        system
+    }
+
+    /// Checks whether a bus interaction can be replaced by a low-degree constraint + range checks.
+    /// Returns None if no replacement is found.
+    fn try_replace_bus_interaction(
+        &self,
+        bus_interaction: &BusInteraction<GroupedExpression<T, V>>,
+    ) -> Option<Replacement<T, V>> {
+        let bus_id = bus_interaction.bus_id.try_to_number()?;
+        if self.bus_interaction_handler.is_stateful(bus_id) {
+            return None;
+        }
+        for input_output_pair in self.possible_input_output_pairs(bus_interaction) {
```

Comment:
Can you use `self.possible_input_output_pairs(bus_interaction).find_map(...)` here?